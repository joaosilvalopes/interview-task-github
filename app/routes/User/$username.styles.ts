import { Link } from '@remix-run/react';
import styled from 'styled-components';
import Spinner from '~/components/Spinner';

export const UserContainer = styled.main` 
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  min-height: calc(100vh - ${props => props.theme.headerHeight});
  color: ${props => props.theme.colors.main};
  background-color: ${props => props.theme.colors.mainBackground};
`;

export const UserHeader = styled.div`
  display: flex;
  align-items: center;
  flex-direction: column;
  width: 100%;
  padding: 1rem;
  background-color: ${props => props.theme.colors.userHeaderBackground};

  ${props => props.theme.media.gteSm} {
    flex-direction: row;
  }
`;

export const UserAvatar = styled.img`
  width: 5rem;
  height: 5rem;
  border-radius: 50%;
  box-shadow: 0px 2px 2px rgba(0, 0, 0, 0.2);

  ${props => props.theme.media.gteSm} {
    margin-right: 1rem;
  }
`;

export const UserName = styled.h1`
  font-size: 2rem;
  color: ${props => props.theme.colors.username};
  margin: 1rem 0;

  ${props => props.theme.media.gteSm} {
    font-size: 2.5rem;
  }

  ${props => props.theme.media.gteMd} {
    font-size: 3rem;
  }
`;

export const UserRepositoriesCount = styled.p`
  font-size: 1.2rem;
  color: ${props => props.theme.colors.userRepositoriesCount};

  ${props => props.theme.media.gteSm} {
    margin-left: auto;
  }
`;

export const UserRepositoriesList = styled.ul`
  display: flex;
  flex-direction: column;
  width: 100%;
  height: 30rem;
`;

export const UserRepository = styled.li`
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  height: 20%;
  width: 100%;
  padding: 0.75rem;
  border-bottom: 1px solid ${props => props.theme.colors.userRepositoryBorder};

  &:first-child {
    border-top: 1px solid ${props => props.theme.colors.userRepositoryBorder};
  }
`;

export const UserRepositoryName = styled.h2`
  font-size: 2rem;
  color: ${props => props.theme.colors.userRepositoryName};
`;

export const UserRepositoryDescription = styled.p`
  font-size: 1.2rem;
  color: ${props => props.theme.colors.userRepositoriesDescription};
  white-space: nowrap;
  overflow-x: clip;
  text-overflow: ellipsis;
`;

export const Pagination = styled.div`
  display: flex;
  width: 100%;
  justify-content: center;
  margin-top: 1rem;
`;

export const PaginationButton = styled.button`
  padding: 0.5rem 1rem;
  font-size: 1.2rem;
  border: none;
  border-radius: 0.5rem;
  color: ${props => props.theme.colors.userPaginationButton};
  background-color: ${props => props.theme.colors.userPaginationButtonBackground};
  cursor: pointer;
  transition: background-color 0.2s ease-in-out;

  &:not(:disabled):hover {
    background-color: ${props => props.theme.colors.userPaginationButtonBackgroundOnHover};
  }

  &:disabled {
    cursor: wait;
  }
`;

export const BackLink = styled(Link)`
  display: block;
  margin: auto 0 1rem 0.75rem;
  padding: 0.5rem 1rem;
  border-radius: 0.5rem;
  cursor: pointer;
  font-size: 1.2rem;
  text-decoration: none;
  transition: background-color 0.2s ease-in-out;
  color: ${props => props.theme.colors.backButton};
  background-color: ${props => props.theme.colors.backButtonBackground};

  &:hover {
    background-color: ${props => props.theme.colors.backButtonBackgroundOnHover};
  }
`;

export const PageCount = styled.span`
  display: flex;
  align-items: center;
  margin: 0 1rem;
  color: ${props => props.theme.colors.pageCount};
  font-size: 1.2rem;
`;

export const PageSpinner = styled(Spinner)`
  font-size: 10rem;
  border-width: 0.5rem;
  margin: auto;
  color: ${props => props.theme.colors.pageSpinner};
`;
