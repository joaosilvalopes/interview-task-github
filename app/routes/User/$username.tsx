import type { LoaderArgs } from '@remix-run/node';
import { useLoaderData } from '@remix-run/react';
import Spinner from '~/components/Spinner';
import usePagination from '~/hooks/usePagination';

import type { Repository } from '~/sdk/getRepositories';
import getRepositories from '~/sdk/getRepositories';
import type { User } from '~/sdk/getUser';
import getUser from '~/sdk/getUser';

import {
  BackLink,
  Pagination,
  PaginationButton,
  UserAvatar,
  UserHeader,
  UserName,
  UserRepositoriesCount,
  UserRepositoriesList,
  UserRepository,
  UserRepositoryDescription,
  UserRepositoryName,
  PageCount,
  UserContainer,
  PageSpinner
} from './$username.styles';

type UserLoaderArgs = LoaderArgs & {
  params: {
    username: string;
  }
};

type ServerData = {
  user: User;
  repositories: Repository[];
};

export const PAGE_SIZE = 5;

export const loader = async ({ params }: UserLoaderArgs) => {
  const { username } = params;

  try {
    const [user, repositories] = await Promise.all([
      getUser(username),
      getRepositories(username, 1, PAGE_SIZE),
    ]);

    return { user, repositories };
  } catch (e) {
    console.error(e);
  }
};

const UserPage = () => {
  const { user, repositories } = useLoaderData<ServerData>();
  const fetchPage = (page: number) => getRepositories(user.username, page, PAGE_SIZE);

  const {
    pageEntries: visibleRepos,
    isPageLoading,
    page,
    pageCount,
    nextPage,
    previousPage
  } = usePagination<Repository>({
    pageSize: PAGE_SIZE,
    firstPageEntries: repositories,
    entryCount: user.repositoriesCount,
    fetchPage
  });

  return (
    <UserContainer>
      <UserHeader>
        <UserAvatar src={user.avatarUrl} data-testid="user-avatar" />
        <UserName data-testid="user-name">{user.name}</UserName>
        <UserRepositoriesCount data-testid="user-repos-count">Found {user.repositoriesCount} repositories</UserRepositoriesCount>
      </UserHeader>
      <UserRepositoriesList>
        {isPageLoading ? <PageSpinner data-testid="loading-page-indicator" /> : visibleRepos.map(repo => (
          <UserRepository key={repo.id}>
            <UserRepositoryName data-testid={`repo-name-${repo.id}`}>{repo.name}</UserRepositoryName>
            {repo.description && <UserRepositoryDescription data-testid={`repo-description-${repo.id}`}>{repo.description}</UserRepositoryDescription>}
          </UserRepository>
        ))}
      </UserRepositoriesList>
      <Pagination>
        {page > 1 && (
          <PaginationButton data-testid="previous-page-button" onClick={previousPage}>
            Previous
          </PaginationButton>
        )}
        <PageCount data-testid="page-count">{page}/{pageCount}</PageCount>
        {page < pageCount && (
          <PaginationButton data-testid="next-page-button" disabled={isPageLoading} onClick={nextPage}>
            { isPageLoading ? <Spinner /> : 'Next' } 
          </PaginationButton>
        )}
      </Pagination>
      <BackLink data-testid="back-link" to="/">Back</BackLink>
    </UserContainer>
  );
};

export default UserPage;
