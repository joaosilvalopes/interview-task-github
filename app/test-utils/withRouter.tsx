import { useState, useEffect, ComponentType } from "react";
import { To, Router, Routes, Route } from "react-router";
import { History } from "@remix-run/router";

const withRouter = (Component: ComponentType, { history, path }: { history: History, path: string }) => () => {
    const [location, setLocation] = useState(history.location);

    useEffect(() => {
        history.listen(({ location }) => setLocation(location));
    }, []);

    const push = (to: To) => setTimeout(() => history.push(to), 100); // simulate remix loader fetch before changing route

    return <Router location={location} navigator={{ ...history, push }}>
        <Routes>
            <Route path={path} element={<Component />} />
            <Route path="*" element={null} />
        </Routes>
    </Router>;
};

export default withRouter;
