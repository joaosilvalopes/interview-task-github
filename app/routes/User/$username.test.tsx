import { createMemoryHistory } from '@remix-run/router';
import { fireEvent, render, waitFor } from '@testing-library/react';

import withRouter from '~/test-utils/withRouter';

import ThemeProvider from '~/theme';

import User, { PAGE_SIZE } from './$username';

const user = {
    username: 'testuser',
    name: 'Test User',
    avatarUrl: 'https://example.com/avatar.png',
    repositoriesCount: PAGE_SIZE * 6
};

const repositories = Array.from({ length: user.repositoriesCount }, (_, i) => ({
    id: i + 1,
    name: `Repository ${i + 1}`,
    description: `Description of Repository ${i + 1}`
}));

const repositoriesFirstPage = repositories.slice(0, PAGE_SIZE);

jest.mock('~/sdk/getRepositories', () => jest.fn((_,page) => {
    return new Promise((resolve) => setTimeout(() => resolve(repositories.slice((page - 1) * PAGE_SIZE, page * PAGE_SIZE)), 100));
  }));

jest.mock('@remix-run/react', () => ({
    ...jest.requireActual('@remix-run/react'),
    useLoaderData: jest.fn(() => ({
        user,
        repositories: repositoriesFirstPage,
    })),
}));

const history = createMemoryHistory({ initialEntries: [`/user/${user.username}`], v5Compat: true });

const UserWithRouter = withRouter(User, { path: '/user/:username', history });

const renderWithRouter = () => render(<ThemeProvider><UserWithRouter /></ThemeProvider>);

describe('UserPage component', () => {
    const expectPageResult = async (page: number, getByTestId: (testId: string) => HTMLElement, getAllByTestId: (testId: RegExp) => HTMLElement[]) => {
        const pageRepositories = repositories.slice(PAGE_SIZE * (page - 1), PAGE_SIZE * page);

        await waitFor(() => expect(getAllByTestId(/^repo-name-[0-9]+/)).toHaveLength(pageRepositories.length));
        await waitFor(() => expect(getAllByTestId(/^repo-description-[0-9]+/)).toHaveLength(pageRepositories.length));

        for(const repo of pageRepositories) {
            await waitFor(() => expect(getByTestId(`repo-name-${repo.id}`)).toHaveTextContent(repo.name));
            await waitFor(() => expect(getByTestId(`repo-description-${repo.id}`)).toHaveTextContent(repo.description));
        }
    }

    it('renders user information and repositories', async () => {
        const { getByTestId, getAllByTestId, queryByTestId, asFragment } = renderWithRouter();

        expect(getByTestId('user-avatar')).toHaveAttribute("src", user.avatarUrl);
        expect(getByTestId('user-name')).toHaveTextContent(user.name);
        expect(getByTestId('user-repos-count')).toHaveTextContent(`Found ${user.repositoriesCount} repositories`);

        const lastPage = Math.ceil(user.repositoriesCount / PAGE_SIZE);

        // First page
        await expectPageResult(1, getByTestId, getAllByTestId);

        expect(queryByTestId(`loading-page-indicator`)).not.toBeInTheDocument();
        expect(queryByTestId(`previous-page-button`)).not.toBeInTheDocument();
        expect(getByTestId(`next-page-button`)).toBeInTheDocument();
        expect(getByTestId(`next-page-button`)).toBeEnabled();
        expect(getByTestId('page-count')).toHaveTextContent(`1/${lastPage}`);

        // Middle Pages
        for(let page = 2; page < lastPage; page++) {
            fireEvent.click(getByTestId('next-page-button'));

            expect(getByTestId(`loading-page-indicator`)).toBeInTheDocument();
            expect(getByTestId(`next-page-button`)).toBeDisabled();

            await expectPageResult(page, getByTestId, getAllByTestId);

            expect(queryByTestId(`loading-page-indicator`)).not.toBeInTheDocument();
            expect(getByTestId(`previous-page-button`)).toBeInTheDocument();
            expect(getByTestId(`next-page-button`)).toBeInTheDocument();
            expect(getByTestId(`next-page-button`)).toBeEnabled();
            expect(getByTestId('page-count')).toHaveTextContent(`${page}/${lastPage}`);
        }

        //Last page
        fireEvent.click(getByTestId('next-page-button'));

        expect(getByTestId(`loading-page-indicator`)).toBeInTheDocument();
        expect(queryByTestId(`next-page-button`)).not.toBeInTheDocument();

        await expectPageResult(lastPage, getByTestId, getAllByTestId);

        expect(queryByTestId(`loading-page-indicator`)).not.toBeInTheDocument();
        expect(getByTestId(`previous-page-button`)).toBeInTheDocument();
        expect(queryByTestId(`next-page-button`)).not.toBeInTheDocument();
        expect(getByTestId('page-count')).toHaveTextContent(`${lastPage}/${lastPage}`);

        expect(asFragment()).toMatchSnapshot();
    });

    it('redirects to root route on back link click', async () => {
        const { asFragment, getByTestId } = renderWithRouter();
        const backLink = getByTestId(`back-link`);

        expect(backLink).toBeInTheDocument();

        fireEvent.click(backLink);

        await waitFor(() => expect(history.location.pathname).toBe('/'));

        expect(asFragment()).toMatchSnapshot();
    });
});
