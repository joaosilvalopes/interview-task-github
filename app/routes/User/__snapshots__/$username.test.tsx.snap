// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPage component redirects to root route on back link click 1`] = `<DocumentFragment />`;

exports[`UserPage component renders user information and repositories 1`] = `
<DocumentFragment>
  <main
    class="sc-guDLRT fLwWYZ"
  >
    <div
      class="sc-dmyDGi kYKMGh"
    >
      <img
        class="sc-hLQRIN cHviBY"
        data-testid="user-avatar"
        src="https://example.com/avatar.png"
      />
      <h1
        class="sc-eDLJxc eZzYKw"
        data-testid="user-name"
      >
        Test User
      </h1>
      <p
        class="sc-jTQDnj jDKHBq"
        data-testid="user-repos-count"
      >
        Found 30 repositories
      </p>
    </div>
    <ul
      class="sc-gLLvby iXzDYo"
    >
      <li
        class="sc-iBdmCd cJQiTP"
      >
        <h2
          class="sc-fsYeqs eIEwIc"
          data-testid="repo-name-26"
        >
          Repository 26
        </h2>
        <p
          class="sc-qZqnM iOLHeu"
          data-testid="repo-description-26"
        >
          Description of Repository 26
        </p>
      </li>
      <li
        class="sc-iBdmCd cJQiTP"
      >
        <h2
          class="sc-fsYeqs eIEwIc"
          data-testid="repo-name-27"
        >
          Repository 27
        </h2>
        <p
          class="sc-qZqnM iOLHeu"
          data-testid="repo-description-27"
        >
          Description of Repository 27
        </p>
      </li>
      <li
        class="sc-iBdmCd cJQiTP"
      >
        <h2
          class="sc-fsYeqs eIEwIc"
          data-testid="repo-name-28"
        >
          Repository 28
        </h2>
        <p
          class="sc-qZqnM iOLHeu"
          data-testid="repo-description-28"
        >
          Description of Repository 28
        </p>
      </li>
      <li
        class="sc-iBdmCd cJQiTP"
      >
        <h2
          class="sc-fsYeqs eIEwIc"
          data-testid="repo-name-29"
        >
          Repository 29
        </h2>
        <p
          class="sc-qZqnM iOLHeu"
          data-testid="repo-description-29"
        >
          Description of Repository 29
        </p>
      </li>
      <li
        class="sc-iBdmCd cJQiTP"
      >
        <h2
          class="sc-fsYeqs eIEwIc"
          data-testid="repo-name-30"
        >
          Repository 30
        </h2>
        <p
          class="sc-qZqnM iOLHeu"
          data-testid="repo-description-30"
        >
          Description of Repository 30
        </p>
      </li>
    </ul>
    <div
      class="sc-jsEegt eoPztV"
    >
      <button
        class="sc-kFCsca kGWwQt"
        data-testid="previous-page-button"
      >
        Previous
      </button>
      <span
        class="sc-csKIKE dKxdqC"
        data-testid="page-count"
      >
        6/6
      </span>
    </div>
    <a
      class="sc-irLwvL elpMmb"
      data-testid="back-link"
      href="/"
    >
      Back
    </a>
  </main>
</DocumentFragment>
`;
